src/cmd/devdraw/cocoa-screen-mojave.m

```diff
@@ -0,0 +1,658 @@
+{
+	NSRect r;
+
+	r = [v rectValue];
+	LOG(@"callsetNeedsDisplayInRect(%g, %g, %g, %g)", r.origin.x, r.origin.y, r.size.width, r.size.height);
+	r = [win convertRectFromBacking:r];
+	LOG(@"setNeedsDisplayInRect(%g, %g, %g, %g)", r.origin.x, r.origin.y, r.size.width, r.size.height);
+	[myContent setNeedsDisplayInRect:r];
+}
+
++ (void)callsetcursor:(NSValue *)v
+{
+	Cursor *c;
+	NSBitmapImageRep *r;
+	NSImage *i;
+	NSPoint p;
+	int b;
+	uchar *plane[5];
+
+	qlock(&cursorl);
+	c = [v pointerValue];
+	if(!c)
+		c = &bigarrow;
+	r = [[NSBitmapImageRep alloc]
+		initWithBitmapDataPlanes:nil
+		pixelsWide:16
+		pixelsHigh:16
+		bitsPerSample:1
+		samplesPerPixel:2
+		hasAlpha:YES
+		isPlanar:YES
+		colorSpaceName:NSDeviceWhiteColorSpace
+		bytesPerRow:2
+		bitsPerPixel:1];
+
+	[r getBitmapDataPlanes:plane];
+
+	for(b=0; b<2*16; b++){
+		plane[0][b] = ~c->set[b];
+		plane[1][b] = c->clr[b];
+	}
+	p = NSMakePoint(-c->offset.x, -c->offset.y);
+	i = [[NSImage alloc] initWithSize:NSMakeSize(16, 16)];
+	[i addRepresentation:r];
+
+	currentCursor = [[NSCursor alloc] initWithImage:i hotSpot:p];
+	qunlock(&cursorl);
+
+	[win invalidateCursorRectsForView:myContent];
+}
+
+- (void)applicationDidFinishLaunching:(id)arg
+{
+	LOG(@"applicationDidFinishLaunching");
+	[NSThread
+		detachNewThreadSelector:@selector(callservep9p:)
+		toTarget:[self class] withObject:nil];
+}
+
+- (NSApplicationPresentationOptions)window:(id)arg
+		willUseFullScreenPresentationOptions:(NSApplicationPresentationOptions)proposedOptions {
+	NSApplicationPresentationOptions o;
+	o = proposedOptions;
+	o &= ~(NSApplicationPresentationAutoHideDock | NSApplicationPresentationAutoHideMenuBar);
+	o |= NSApplicationPresentationHideDock | NSApplicationPresentationHideMenuBar;
+	return o;
+}
+
+- (void)windowDidEndLiveResize:(id)arg
+{
+	resizeimg();
+}
+
+- (void)applicationWillTerminate:(NSNotification *)aNotification {
+}
+
+- (BOOL)applicationShouldTerminateAfterLastWindowClosed:(NSApplication *)theApplication {
+	return YES;
+}
+
+@end
+
+@implementation DevDrawView
+
+- (id)initWithFrame:(NSRect)r
+{
+	self = [super initWithFrame:r];
+	LOG(@"initWithFrame(%g,%g,%g,%g)", r.origin.x, r.origin.y, r.size.width, r.size.height);
+	return self;
+}
+
+- (void)drawRect:(NSRect)r
+{
+	NSRect sr, dr;
+	CGContextRef c;
+	CGImageRef i;
+	int l;
+
+	if(!img){
+		LOG(@"img not initialized");
+		return;
+	}
+
+	c = [[NSGraphicsContext currentContext] CGContext];
+
+	LOG(@"drawRect(%g,%g,%g,%g)", r.origin.x, r.origin.y, r.size.width, r.size.height);
+	sr = [win convertRectToBacking:r];
+	LOG(@"sr: %g,%g,%g,%g", sr.origin.x, sr.origin.y, sr.size.width, sr.size.height);
+	dr = r;
+	dr.origin.y = self.bounds.size.height - dr.origin.y - dr.size.height;
+	LOG(@"dr: %g,%g,%g,%g", dr.origin.x, dr.origin.y, dr.size.width, dr.size.height);
+
+	l = img->width*sizeof(u32int);
+	CGDataProviderRef provider = CGDataProviderCreateWithData(NULL,
+		byteaddr(img, Pt(sr.origin.x, sr.origin.y)), sr.size.height*l, NULL);
```

Comment:
For my understanding, why is this thread-safe? Doesn't `img` get written to by the `servep9p` thread?

Reply:
No.  It's not thread-safe.

I was relying on the fact the drawRect will always be called after the devdraw thread modifies the memscreen and calls _flushmemscreen.  Even if the read/write of two threads collide, and drawRect reads and draws garbage, it will be called again.

I will try a thread safe implementation, similar to what you did.

Reply:
Thanks, that makes sense. I guess you could call it "eventually consistent" drawing :)